### PR TITLE
Added Event "VisibleTilesDidLoad" published after visible page tiles are drawn

### LINKED ISSUE #386: Trigger an Event after drawing tiles of a page
This is useful for plugins who want to draw on top of a page in Diva.

### PR DESCRIPTION
**Pull request by zeyadhazem**
This will be used by plugins who want to draw on top of a page in Diva.
Fixes #386

### CODE REVIEW

File: source/js/renderer.js [hunk 1]
```diff
@@ -219,6 +226,7 @@ export default class Renderer
                     this._drawTile(pageIndex, scaled, this._cache.get(source.url));
                 }
             }, this);
+            this.publish("VisibleTilesDidLoad", pageIndex, this._zoomLevel);
```
Comment (ahankinson) [inline source/js/renderer.js]: You should instead check out the hooks that are passed into this class when it's instantiated. Then you don't need to import the diva instance here, and you get access to the publish method that's in the viewer core.

https://github.com/DDMAL/diva.js/blob/develop-diva6/source/js/viewer-core.js#L484-L506

**Comment by ahankinson** [comment]
How many times on a single image load does this event fire?

**Comment by zeyadhazem** [comment]
A lot of times at the first load, but 2 times after that (when scrolling, etc...). Maybe it shouldn't be VisibleTilesDidLoad but TilesStreamDidLoad to be more accurate?

File: source/js/renderer.js (hunk 1 above)
Comment (zeyadhazem) [inline source/js/renderer.js]: This is EXACTLY what I was looking for! Super awesome!! Thanks

**Comment by ahankinson** [comment]
But if it fires lots of times, then (I presume) your highlight drawing code will start drawing on the *first* call, and that may be before all the tiles are loaded. This is what I keep trying to say -- you can't just put your call at the end of the loop, because you're firing off a whole bunch of async processes.

Why would you rename the event?

**Comment by zeyadhazem** [comment]
No the highlight code will not start drawing just on the first call, it will keep redrawing every time new tiles are being painted. Every time the async process fires, there will be redrawing of the visible tiles of a whole page on the screen and after that the highlights will start being drawn. This can be done several times, no problem there will be redundancy and highlighting the same things again but it will always work!
I have already tested this and it works fine. Unless you have other extreme test cases that you would like me to try.

**Comment by ahankinson** [comment]
Tile 1 returns. 
Loop through all returned tiles
Paint tile 1
Tile 2 returns
Loop through all returned tiles
Paint tile 1 and tile 2
Tile 3 returns
Loop through all returned tiles
Paint tile 1, 2, and 3.
...
Tile n returns
Loop through all returned tiles
Paint tile1, 2, 3, ... n

So you are painting n! (factorial) (n + n-1 + n-2) times. That seems like a sub-optimal solution to me.

**Comment by zeyadhazem** [comment]
Yes but this is only happening at the beginning, everything after the first page is drawn is being drawn in O(n), I will look into it and get some traces next week, but for now this works for Pixel.

**Comment by ahankinson** [comment]
But you haven't actually fixed it, so I can't merge it.

**Comment by zeyadhazem** [comment]
Dropping this PR!

**Pull request closed without merging**